HW0/P4/P4.py

```diff
@@ -23,55 +23,112 @@
     # Part 2:
     #
     # Read the observation array and plot it (Part 2)
     #####################
-
-    # ax.plot(x_coords, y_coords, z_coords,
-    #         '.g', label='Observed trajectory')
-
+    
+    meas=np.loadtxt("P4_measurements.txt", delimiter=",");
+    np.size(meas)
+    np.reshape(meas,(K,3));
+    meas=np.transpose(meas)
+
+    invr_vec = np.array([1./rx,1./ry,1./rz])
+    invC= np.diag(invr_vec)
+    m=np.dot(invC,meas)
+
+    x_coords,y_coords,z_coords=m[0:3,:];
+    ax.plot(x_coords, y_coords, z_coords, '.g', label='Observed trajectory')
+    
     #####################
     # Part 3:
     # Use the initial conditions and propagation matrix for prediction
     #####################
+    
+    A = np.eye(6)+np.diag(dt*np.ones(3),3)-np.diag(np.concatenate((np.zeros(3), c*dt*np.ones(3))))
+    a = np.concatenate((np.zeros(5),np.array([g*dt])))
+    s = []
 
-    # A = ?
-    # a = ?
-    # s = ?
 
     # Initial conditions for s0
+    s0= np.array([0., 0., 2., 15., 3.5, 4.0])
+    s.append(s0)
+    
     # Compute the rest of sk using Eq (1)
-
-    # ax.plot(x_coords, y_coords, z_coords,
-    #         '-k', label='Blind trajectory')
-
+    for k in range(1,K,1):
+        sk = np.dot(A,s[-1])+a
+        s.append(sk)
+    s=np.array(s)
+    np.reshape(s,(6,K))
+    s=np.transpose(s)
+    
+    x_coords,y_coords,z_coords=s[0:3,:]
+    ax.plot(x_coords, y_coords, z_coords,
+            '-k', label='Blind trajectory')
+    
     #####################
     # Part 4:
     # Use the Kalman filter for prediction
     #####################
 
-    # B = ?
-    # C = ?
+    B = np.diag(np.array([bx,by,bz,bvx,bvy,bvz]))
+    C = np.zeros((3,6))
+    C[:3,:3] = np.diag(np.array([rx,ry,rz]))
+    
+    def prefictS(s):
+        return np.dot(A,s)+a
+    
+    def predictSig(sig):
+        return np.linalg.inv(np.dot(np.dot(A,sig),np.transpose(A))+np.dot(B,np.transpose(B)))
+    
+    def updateSig(sig):
+        return np.linalg.inv(sig+np.dot(np.transpose(C),C))
+    
+    def updateS(sigk,sig,s,m):
+        return np.dot(sigk,(np.dot(sig,s)+np.dot(np.transpose(C),m)))
+
 
     # Initial conditions for s0 and Sigma0
+    s2 = []
+    s0= np.array([0., 0., 2., 15., 3.5, 4.0])
+    s2.append(s0)
+    Sigma0=np.eye(6)*.01
+    
     # Compute the rest of sk using Eqs (2), (3), (4), and (5)
-
-    # ax.plot(x_coords, y_coords, z_coords,
-    #         '-r', label='Filtered trajectory')
-
+    curs = prefictS(s0)
+    curSig = predictSig(Sigma0)
+    curSigk = updateSig(curSig)
+    
+    for k in range(1,K,1):
+        s2.append(updateS(curSigk,curSig,curs,meas[:,k]))
+        curs = prefictS(s2[-1])
+        curSig = predictSig(curSigk)
+        curSigk = updateSig(curSig)
+    s2=np.array(s2)
+    np.reshape(s2,(6,K))
+    s2=np.transpose(s2)
+
+    x_coords, y_coords, z_coords = s2[0:3,:]
+    ax.plot(x_coords, y_coords, z_coords,
+            '-r', label='Filtered trajectory')
+    
     # Show the plot
     ax.legend()
     plt.show()
```

Comment:
I really can't make an intelligent comment on your code.  This assignment was the first thing I ever did in Python, so it is really amazing to look at the work of someone who actually knew what they were doing.  Lots of functions that I didn't know about, and all of the slicing and matrix operations were done very elegantly and succinctly.